File: 0169-majority-element/0169-majority-element.py
# Boyer-Moore Voting Algorithm
# This method works in O(n) time and O(1) space.
# Thank you ChatGPT. (-:

class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        candidate, count = None, 0

        for num in nums:
            if count == 0:
                candidate = num
            count += 1 if num == candidate else -1
        
        return candidate

File: 0169-majority-element/0169-majority-element.rs
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut candidate: i32 = 0;
        let mut count: i32 = 0;

        for num in nums.into_iter() {
            if count == 0 {
                candidate = num;
            }
            count += if num == candidate { 1 } else { -1 };
        }

        candidate
    }
}